utilix/mongo_files.py

```diff
@@ -0,0 +1,574 @@
+        for config in self.list_files():
+            self.download_single(config)
+
+    @staticmethod
+    def _check_store_files_at(cache_folder_alternatives):
+        """
+        Iterate over the options in cache_options until we find a folder
+            where we can store data. Order does matter as we iterate
+            until we find one folder that is willing.
+
+        :param cache_folder_alternatives: tuple, this tuple must be a
+            list of paths one can try to store the downloaded data
+
+        :return: str, the folder that we can write to.
+        """
+        if not isinstance(cache_folder_alternatives, (tuple, list)):
+            raise ValueError('cache_folder_alternatives must be tuple')
+        for folder in cache_folder_alternatives:
+            if not os.path.exists(folder):
+                try:
+                    os.makedirs(folder)
+                except (PermissionError, OSError):
+                    continue
+            if os.access(folder, os.W_OK):
+                return folder
+        raise PermissionError(
+            f'Cannot write to any of the cache_folder_alternatives: '
+            f'{cache_folder_alternatives}')
+
+
+class DownloadAPI(MongoDownloader):
+    def __init__(self,
+                 readonly=True,
+                 config_identifier='config_name',
+                 ):
+        self.DB = utilix.DB()
+        self.readonly = readonly
+        self.config_identifier = config_identifier
+
+    def config_exists(self, config):
+        """
+        Quick check if this config is already saved in the collection
+
+        :param config: str,  name of the file of interest
+        :return: bool, is this config name stored in the database
+        """
+        query = self.get_query_config(config)
+        return self.DB.count_documents(query) > 0
+
+    def md5_stored(self, abs_path):
+        """
+        NB: RAM intensive operation!
+        Carefully compare if the MD5 identifier is the same as the file
+        as stored under abs_path.
+
+        :param abs_path: str, absolute path to the file name
+        :return: bool, returns if the exact same file is already stored
+            in the database
+
+        """
+        if not os.path.exists(abs_path):
+            # A file that does not exist does not have the same MD5
+            return False
+        query = {'md5': self.compute_md5(abs_path)}
+        return self.DB.count_documents(query) > 0
+
+    def list_files(self):
+        """
+        Get a complete list of files that are stored in the database
+
+        :return: list, list of the names of the items stored in this
+            database
+
+        """
+        return [doc[self.config_identifier]
+                for doc in
+                self.DB.file_find(
+                    projection=
+                    {self.config_identifier: 1})
+                if self.config_identifier in doc
+                ]
+
+    def get_gridfs_object(self,

```

Comment:
Hey Evan, this sounds like you are not convinced we need this PR at all with this class to take care of checking the MD5, preventing simultaneous read/writes, checking if a file is already downloaded, checking if a file should be uploaded in the first place etc etc. If this is the case, we should have a chat because then we are on a very different page.

Otherwise, if indeed getting the gridf is hard, yes, your approach makes sense.